### PR TITLE
NIAD-348: MESH authorization token

### PR DESCRIPTION
**Pull request by grzegorz-robak**

### CODE REVIEW

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/AuthorizationHashGenerator.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.apache.commons.codec.binary.Hex;
+
+class AuthorizationHashGenerator {
+
+    @SneakyThrows
+    public String computeHash(MeshConfig meshConfig, Nonce nonce, String timestamp) {
+        String hashInput = String.join(":", meshConfig.getMailboxId(), nonce.value, nonce.count, meshConfig.getMailboxPassword(), timestamp);
+
+        Mac sha256_HMAC = Mac.getInstance("HmacSHA256");
```
Comment (jaroslaw-wojciechowski) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/AuthorizationHashGenerator.java]: HmacSHA256 should be static

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/MeshAuthorizationToken.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    private final String hash;
+
+    public MeshAuthorizationToken(MeshConfig meshConfig, Instant timestamp, Nonce nonce, AuthorizationHashGenerator authorizationHashGenerator) {
+        String prefix = "NHSMESH " + meshConfig.getMailboxId();
+        String currentTimeFormatted = new TokenTimestamp(timestamp).getValue();
+        this.data = String.join(":", prefix, nonce.value, nonce.count, currentTimeFormatted);
+        this.hash = authorizationHashGenerator.computeHash(meshConfig, nonce, currentTimeFormatted);
+    }
+
+    public MeshAuthorizationToken(MeshConfig meshConfig) {
```
Comment (jaroslaw-wojciechowski) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/MeshAuthorizationToken.java:26]: MeshAuthorizationToken not used?

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/MeshAuthorizationToken.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * MESH authorization token
+ * One time use only - each MESH API call should use new token
+ */
+public class MeshAuthorizationToken {
+
+    private final String data;
+    private final String hash;
+
+    public MeshAuthorizationToken(MeshConfig meshConfig, Instant timestamp, Nonce nonce, AuthorizationHashGenerator authorizationHashGenerator) {
+        String prefix = "NHSMESH " + meshConfig.getMailboxId();
```
Comment (jaroslaw-wojciechowski) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/MeshAuthorizationToken.java]: NHSMESH should be static

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/AuthorizationHashGeneratorTest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package uk.nhs.digital.nhsconnect.nhais.mesh.token;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneId;
```
Comment (jaroslaw-wojciechowski) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/AuthorizationHashGeneratorTest.java]: 2 unused imports

**Review by jaroslaw-wojciechowski — CHANGES_REQUESTED** [review]

File: src/test/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/MeshAuthorizationTokenTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package uk.nhs.digital.nhsconnect.nhais.mesh.token;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneId;
```
Comment (jaroslaw-wojciechowski) [inline src/test/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/MeshAuthorizationTokenTest.java]: unused imports

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/MeshAuthorizationToken.java (hunk 2 above)
Comment (grzegorz-robak) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/token/MeshAuthorizationToken.java:26]: As of now it isn't - I'll use it in integration test

**Review by jaroslaw-wojciechowski — APPROVED** [review]

**Review by Alicja-Kainos — APPROVED** [review]

**Pull request merged**